src/Streamly/Mem/Array.hs

```diff
@@ -321,27 +385,57 @@ readIndex arr i =
     else liftIO $ withForeignPtr (aStart arr) $ \p ->
             fmap Just $ peekElemOff p i
 
-{-
+
 -- | @readSlice arr i count@ streams a slice of the array @arr@ starting
 -- at index @i@ and reading up to @count@ elements in the forward direction
 -- ending at the index @i + count - 1@.
 --
+-- >>> S.toList $ readSlice (fromList [1..10]) 3 4
+-- > [4,5,6,7]
+--
 -- @since 0.7.0
 {-# INLINE readSlice #-}
-readSlice :: (IsStream t, Monad m, Storable a)
+readSlice :: (IsStream t, MonadIO m, Storable a)
     => Array a -> Int -> Int -> t m a
-readSlice arr i len = undefined
+readSlice arr i len = D.fromStreamD (D.Stream step i)
+  where
+    step _ j
+        | j < i + len = do
+            r <- liftIO $ readIndex arr j
```

Comment:
`readIndex` performs a bounds check every time and then we are checking for the return value for Nothing or Just. We can avoid that by reading directly instead of using `readIndex`, that will be more efficient.